File: app-android/app/src/main/java/com/remotecontrol/app/net/Protocol.kt
package com.remotecontrol.app.net

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okio.ByteString

const val PROTOCOL_VERSION = 6

val ProtoJson = Json {
    ignoreUnknownKeys = true
    classDiscriminator = "type"
    encodeDefaults = true
}

// ---------- Client -> Server ----------

@Serializable
sealed interface ClientMsg

@Serializable
@SerialName("hello")
data class Hello(
    val v: Int = PROTOCOL_VERSION,
    val c: String,
    val nonce: String,
    val client: ClientInfo,
) : ClientMsg

/**
 * Skip-the-QR reconnect handshake. Sent in place of [Hello] when we
 * already have a `(deviceId, token)` minted from a previous successful
 * pairing on this server. Server verifies against its trusted-devices
 * file; on match we get back the same Welcome shape (minus the new
 * trust_token / device_id fields, which we're already holding).
 */
@Serializable
@SerialName("trusted_hello")
data class TrustedHello(
    val v: Int = PROTOCOL_VERSION,
    @SerialName("device_id") val deviceId: String,
    val token: String,
    val client: ClientInfo,
) : ClientMsg

@Serializable
@SerialName("ping")
data class Ping(val ts: Long) : ClientMsg

@Serializable
@SerialName("stream_request")
data class StreamRequest(
    val codec: String = "h264",
    @SerialName("max_bitrate_kbps") val maxBitrateKbps: Int? = null,
    @SerialName("max_fps") val maxFps: Int? = null,
    @SerialName("prefer_keyframe_interval_ms") val preferKeyframeIntervalMs: Int? = null,
) : ClientMsg

@Serializable
@SerialName("stream_stop")
data class StreamStop(@SerialName("stream_id") val streamId: String? = null) : ClientMsg

@Serializable
@SerialName("keyframe_request")
data class KeyframeRequest(@SerialName("stream_id") val streamId: String? = null) : ClientMsg

// ---- M3 mouse input ----

/** Mouse position normalized to [0.0, 1.0]. */
@Serializable
@SerialName("mouse_move")
data class MouseMove(val x: Float, val y: Float) : ClientMsg

@Serializable
@SerialName("mouse_button")
data class MouseButton(val button: String, val down: Boolean) : ClientMsg

/** Wheel deltas in notches; positive = up/right. */
@Serializable
@SerialName("mouse_scroll")
data class MouseScroll(val dx: Int = 0, val dy: Int = 0) : ClientMsg

enum class MouseBtn(val wire: String) {
    Left("left"), Right("right"), Middle("middle"),
}

// ---- M3.5 keyboard input ----

/** Inject Unicode text (works for CJK; bypasses PC IME). */
@Serializable
@SerialName("key_text")
data class KeyText(val text: String) : ClientMsg

/** Win32 virtual-key down/up. See VKey for common codes. */
@Serializable
@SerialName("key_event")
data class KeyEvent(val vk: Int, val down: Boolean) : ClientMsg

// ---- M8 clipboard ----

@Serializable
@SerialName("clipboard_set")
data class ClipboardSet(val text: String) : ClientMsg

@Serializable
@SerialName("clipboard_get")
class ClipboardGet : ClientMsg {
    // Empty payload data class can't be a `data class` without fields, use object-like.
    override fun equals(other: Any?): Boolean = other is ClipboardGet
    override fun hashCode(): Int = 0
}

// ---- M6 file transfer (phone → PC) ----

@Serializable
@SerialName("file_transfer_begin")
data class FileTransferBegin(
    val id: Int,
    val name: String,
    val size: Long,
) : ClientMsg

@Serializable
@SerialName("file_transfer_abort")
data class FileTransferAbort(
    val id: Int,
    val reason: String = "",
) : ClientMsg

// ---- M6 v2 file transfer (PC → phone). Phone is the receiver. ----

@Serializable
@SerialName("file_send_accepted")
data class FileSendAccepted(
    val id: Int,
    @SerialName("dest_path") val destPath: String,
) : ClientMsg

@Serializable
@SerialName("file_send_complete")
data class FileSendComplete(
    val id: Int,
    @SerialName("dest_path") val destPath: String,
) : ClientMsg

@Serializable
@SerialName("file_send_failed")
data class FileSendFailed(
    val id: Int,
    val reason: String,
) : ClientMsg

/** Win32 VK_* constants we expose in the keyboard overlay. */
object VKey {
    const val ESCAPE = 0x1B
    const val TAB = 0x09
    const val BACK = 0x08
    const val RETURN = 0x0D
    const val LEFT = 0x25
    const val UP = 0x26
    const val RIGHT = 0x27
    const val DOWN = 0x28
    const val HOME = 0x24
    const val END = 0x23
    const val PRIOR = 0x21 // PageUp
    const val NEXT = 0x22 // PageDown
    const val SNAPSHOT = 0x2C // PrtSc
    const val LWIN = 0x5B
    const val CONTROL = 0x11
    const val MENU = 0x12 // Alt
    const val SHIFT = 0x10
    const val DELETE = 0x2E
    fun f(n: Int): Int = 0x70 + (n - 1) // F1=0x70 ... F12=0x7B
    /** Letter VK code: 'A' (0x41) ... 'Z' (0x5A). Pass uppercase ASCII. */
    fun letter(c: Char): Int = c.uppercaseChar().code
}

/** Steps inside a [Macro], played sequentially on the server. */
sealed interface MacroStep {
    data class KeyDown(val vk: Int) : MacroStep
    data class KeyUp(val vk: Int) : MacroStep
    /** Convenience: down + up. */
    data class KeyTap(val vk: Int) : MacroStep
    /** Inter-step delay so the PC OS observes events as separate keystrokes. */
    data class Delay(val ms: Long) : MacroStep
}

data class Macro(val label: String, val steps: List<MacroStep>)

/** Built-in shortcut macros (M7 v1). User-defined macros come later. */
object Macros {
    private fun combo(modifier: Int, vk: Int) = listOf(
        MacroStep.KeyDown(modifier),
        MacroStep.Delay(8),
        MacroStep.KeyTap(vk),
        MacroStep.Delay(8),
        MacroStep.KeyUp(modifier),
    )

    val CTRL_C = Macro("Ctrl+C", combo(VKey.CONTROL, VKey.letter('C')))
    val CTRL_V = Macro("Ctrl+V", combo(VKey.CONTROL, VKey.letter('V')))
    val CTRL_X = Macro("Ctrl+X", combo(VKey.CONTROL, VKey.letter('X')))
    val CTRL_A = Macro("Ctrl+A", combo(VKey.CONTROL, VKey.letter('A')))
    val CTRL_Z = Macro("Ctrl+Z", combo(VKey.CONTROL, VKey.letter('Z')))
    val ALT_TAB = Macro("Alt+Tab", combo(VKey.MENU, VKey.TAB))
    val WIN_R = Macro("Win+R", combo(VKey.LWIN, VKey.letter('R')))
    val WIN_D = Macro("Win+D", combo(VKey.LWIN, VKey.letter('D')))

    val DEFAULTS: List<Macro> = listOf(CTRL_C, CTRL_V, CTRL_X, CTRL_A, CTRL_Z, ALT_TAB, WIN_R, WIN_D)
}

@Serializable
data class ClientInfo(
    val name: String,
    val os: String,
    @SerialName("app_version") val appVersion: String,
)

// ---------- Server -> Client ----------

@Serializable
sealed interface ServerMsg

@Serializable
@SerialName("welcome")
data class Welcome(
    val session: String,
    val server: ServerInfo,
    val hmac: String,
    /** Long-lived token to persist for [TrustedHello] reconnects. Null
     *  when this Welcome is itself the response to a TrustedHello (the
     *  client already has the token). */
    @SerialName("trust_token") val trustToken: String? = null,
    /** Stable id the server assigned to this device. Echo back in
     *  [TrustedHello.deviceId]. Null on TrustedHello replies. */
    @SerialName("device_id") val deviceId: String? = null,
) : ServerMsg

@Serializable
@SerialName("error")
data class ServerError(
    val code: String,
    val msg: String,
) : ServerMsg

@Serializable
@SerialName("pong")
data class Pong(val ts: Long) : ServerMsg

@Serializable
@SerialName("stream_started")
data class StreamStarted(
    @SerialName("stream_id") val streamId: String,
    val codec: String,
    val profile: String,
    val level: String,
    val width: Int,
    val height: Int,
    val fps: Int,
    @SerialName("bitrate_kbps") val bitrateKbps: Int,
    @SerialName("keyframe_interval_frames") val keyframeIntervalFrames: Int,
    @SerialName("pixel_format") val pixelFormat: String,
    @SerialName("started_at_unix_ms") val startedAtUnixMs: Long,
    val audio: AudioMetadataDto? = null,
) : ServerMsg

@Serializable
data class AudioMetadataDto(
    val codec: String,
    @SerialName("sample_rate") val sampleRate: Int,
    val channels: Int,
    @SerialName("frame_size_ms") val frameSizeMs: Int,
    @SerialName("bitrate_kbps") val bitrateKbps: Int,
    @SerialName("csd_0_b64") val csd0B64: String,
    @SerialName("csd_1_b64") val csd1B64: String,
    @SerialName("csd_2_b64") val csd2B64: String,
)

@Serializable
@SerialName("stream_stopped")
data class StreamStopped(
    @SerialName("stream_id") val streamId: String,
    val reason: String,
    val msg: String = "",
) : ServerMsg

@Serializable
@SerialName("clipboard_text")
data class ClipboardText(val text: String) : ServerMsg

// ---- M6 server replies ----

@Serializable
@SerialName("file_transfer_accepted")
data class FileTransferAccepted(
    val id: Int,
    @SerialName("dest_path") val destPath: String,
) : ServerMsg

@Serializable
@SerialName("file_transfer_complete")
data class FileTransferComplete(
    val id: Int,
    @SerialName("dest_path") val destPath: String,
) : ServerMsg

@Serializable
@SerialName("file_transfer_failed")
data class FileTransferFailed(
    val id: Int,
    val reason: String,
) : ServerMsg

// ---- M6 v2 inbound announcement from PC ----

@Serializable
@SerialName("file_send_begin")
data class FileSendBegin(
    val id: Int,
    val name: String,
    val size: Long,
) : ServerMsg

@Serializable
data class ServerInfo(
    val name: String,
    val os: String,
    val version: String,
)

// ---------- Binary video frame (data plane) ----------
//
// Wire format (see docs/PROTOCOL.md §"数据平面"):
//   byte 0   : frame_type (0x01 = video)
//   byte 1   : flags     (bit0=keyframe, bit1=config-inlined SPS/PPS)
//   bytes 2-3: reserved (must be 0)
//   bytes 4-11: pts_us (LE u64)
//   bytes 12+: payload — H.264 Annex-B NAL units

object FrameType {
    const val VIDEO = 0x01
    const val AUDIO = 0x02 // reserved for M5
    /** M6 file-transfer chunk. Reuses the 12-byte header layout; the
     *  8 PTS bytes are repurposed as
     *  `transfer_id (LE u32) || chunk_seq (LE u32)`. */
    const val FILE = 0x03
}

object FrameFlags {
    const val KEYFRAME = 0x01
    const val CONFIG = 0x02
    /** M6 file-transfer: marks the FINAL chunk in a transfer. */
    const val LAST_CHUNK = 0x01
}

/**
 * Build a single file-chunk WS Binary frame. Header is 12 bytes; layout
 * mirrors the spec:
 *   byte 0     : frame_type = 0x03
 *   byte 1     : flags     (bit0 = LAST_CHUNK)
 *   bytes 2-3  : reserved (must be 0)
 *   bytes 4-7  : transfer_id (LE u32)
 *   bytes 8-11 : chunk_seq   (LE u32)
 *   bytes 12+  : payload
 *
 * The split of the 8 PTS bytes into two LE u32 fields keeps the header
 * size the same as video / audio frames so transports that already know
 * how to relay a 12-byte-prefixed binary frame don't need any changes.
 */
fun buildFileChunkFrame(
    transferId: Int,
    chunkSeq: Int,
    last: Boolean,
    payload: ByteArray,
): ByteArray {
    val out = ByteArray(FRAME_HEADER_LEN + payload.size)
    out[0] = FrameType.FILE.toByte()
    out[1] = (if (last) FrameFlags.LAST_CHUNK else 0).toByte()
    // bytes 2,3: reserved (0)
    // transfer_id LE u32 at bytes 4-7
    out[4] = (transferId and 0xFF).toByte()
    out[5] = ((transferId ushr 8) and 0xFF).toByte()
    out[6] = ((transferId ushr 16) and 0xFF).toByte()
    out[7] = ((transferId ushr 24) and 0xFF).toByte()
    // chunk_seq LE u32 at bytes 8-11
    out[8] = (chunkSeq and 0xFF).toByte()
    out[9] = ((chunkSeq ushr 8) and 0xFF).toByte()
    out[10] = ((chunkSeq ushr 16) and 0xFF).toByte()
    out[11] = ((chunkSeq ushr 24) and 0xFF).toByte()
    System.arraycopy(payload, 0, out, FRAME_HEADER_LEN, payload.size)
    return out
}

const val FRAME_HEADER_LEN = 12

/** One Opus packet handed to MediaCodec. PTS is on the same monotonic clock as video. */
data class AudioFrame(
    val payload: ByteArray,
    val ptsUs: Long,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AudioFrame) return false
        return ptsUs == other.ptsUs && payload.contentEquals(other.payload)
    }
    override fun hashCode(): Int = (ptsUs.hashCode() * 31 + payload.contentHashCode())
}

data class VideoFrame(
    val type: Int,
    val isKeyframe: Boolean,
    val hasConfig: Boolean,
    val ptsUs: Long,
    val payload: ByteArray,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is VideoFrame) return false
        return type == other.type && isKeyframe == other.isKeyframe &&
                hasConfig == other.hasConfig && ptsUs == other.ptsUs &&
                payload.contentEquals(other.payload)
    }
    override fun hashCode(): Int = (((type * 31 + ptsUs.hashCode()) * 31 +
            payload.contentHashCode()))
}

object FrameParser {
    /**
     * Parse one binary WebSocket frame into a VideoFrame, or return null if
     * malformed (too short, unknown type, reserved bits non-zero — server bug).
     */
    fun parse(bytes: ByteString): VideoFrame? {
        if (bytes.size < FRAME_HEADER_LEN) return null
        val type = bytes[0].toInt() and 0xFF
        val flags = bytes[1].toInt() and 0xFF
        // bytes 2-3 reserved; we don't validate (forward-compat).
        val ptsUs = readLongLe(bytes, 4)
        val payload = bytes.substring(FRAME_HEADER_LEN).toByteArray()
        return VideoFrame(
            type = type,
            isKeyframe = (flags and FrameFlags.KEYFRAME) != 0,
            hasConfig = (flags and FrameFlags.CONFIG) != 0,
            ptsUs = ptsUs,
            payload = payload,
        )
    }

    private fun readLongLe(b: ByteString, off: Int): Long {
        var v = 0L
        for (i in 0 until 8) {
            v = v or ((b[off + i].toLong() and 0xFFL) shl (i * 8))
        }
        return v
    }
}

File: server-pc/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[serde(tag = "type")]
pub enum ClientMsg {
    #[serde(rename = "hello")]
    Hello {
        v: u32,
        c: String,
        nonce: String,
        #[serde(default)]
        client: ClientInfo,
    },

    /// "I've paired with this server before, here's the long-lived token I
    /// got back then." Lets the phone reconnect without scanning a QR code
    /// every time the server restarts. The token is a 256-bit random value
    /// minted by the server on the first successful Hello and remembered
    /// in `trusted_devices.json`.
    #[serde(rename = "trusted_hello")]
    TrustedHello {
        v: u32,
        device_id: String,
        token: String,
        #[serde(default)]
        client: ClientInfo,
    },
    #[serde(rename = "ping")]
    Ping {
        #[serde(default)]
        ts: u64,
    },

    // ---- M2: stream control ----

    #[serde(rename = "stream_request")]
    StreamRequest {
        #[serde(default = "default_codec")]
        codec: String,
        #[serde(default)]
        max_bitrate_kbps: Option<u32>,
        #[serde(default)]
        max_fps: Option<u32>,
        #[serde(default)]
        prefer_keyframe_interval_ms: Option<u32>,
    },

    #[serde(rename = "stream_stop")]
    StreamStop {
        #[serde(default)]
        stream_id: Option<String>,
    },

    #[serde(rename = "keyframe_request")]
    KeyframeRequest {
        #[serde(default)]
        stream_id: Option<String>,
    },

    // ---- M3: mouse input (client → server) ----

    /// Absolute mouse position; x/y normalized to [0.0, 1.0].
    #[serde(rename = "mouse_move")]
    MouseMove { x: f32, y: f32 },

    /// Mouse button press/release.
    #[serde(rename = "mouse_button")]
    MouseButton { button: MouseButton, down: bool },

    /// Mouse wheel delta in notches (positive = up/right, negative = down/left).
    #[serde(rename = "mouse_scroll")]
    MouseScroll {
        #[serde(default)]
        dx: i32,
        #[serde(default)]
        dy: i32,
    },

    // ---- M3.5: keyboard input ----

    /// Unicode text injection (one key event per scalar, KEYEVENTF_UNICODE).
    #[serde(rename = "key_text")]
    KeyText { text: String },

    /// Virtual-key down/up. `vk` is a Win32 VK_* code.
    #[serde(rename = "key_event")]
    KeyEvent { vk: u32, down: bool },

    // ---- M8: clipboard sync ----

    /// Push text from the phone to the PC clipboard.
    #[serde(rename = "clipboard_set")]
    ClipboardSet { text: String },

    /// Request the PC clipboard (server replies with `clipboard_text`).
    #[serde(rename = "clipboard_get")]
    ClipboardGet,

    // ---- M6: file transfer (phone → PC v1) ----

    /// Announce a new file-transfer session. Server allocates an output
    /// path under the user's Downloads/RemoteControl folder, opens the
    /// file, and replies with `FileTransferAccepted`. Subsequent Binary
    /// frames with `frame_type=FILE` and the same `id` carry chunked
    /// content; the final chunk has the `LAST` flag set.
    #[serde(rename = "file_transfer_begin")]
    FileTransferBegin {
        id: u32,
        name: String,
        size: u64,
    },

    /// Voluntary cancel before reaching the LAST chunk. Server closes
    /// and unlinks the partial output.
    #[serde(rename = "file_transfer_abort")]
    FileTransferAbort {
        id: u32,
        #[serde(default)]
        reason: String,
    },

    // ---- M6 v2: file transfer (PC → phone). Phone is the receiver. ----

    /// Phone confirms it opened a destination file for the incoming
    /// transfer announced by `FileSendBegin`. PC may now start
    /// streaming chunks (FILE binary frames carrying the same `id`).
    /// `dest_path` is whatever the phone chose to display to the user —
    /// app-private external storage path, MediaStore URI string, etc.
    #[serde(rename = "file_send_accepted")]
    FileSendAccepted { id: u32, dest_path: String },

    /// Phone has written the last chunk to disk and finalized the file.
    /// Returns the same `dest_path` so the PC UI can confirm.
    #[serde(rename = "file_send_complete")]
    FileSendComplete { id: u32, dest_path: String },

    /// Phone aborted: out of space, IO error, user declined, transfer
    /// id unknown, etc. PC reports the reason and forgets this id.
    #[serde(rename = "file_send_failed")]
    FileSendFailed { id: u32, reason: String },
}

#[derive(Debug, Deserialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MouseButton {
    Left,
    Right,
    Middle,
}

fn default_codec() -> String {
    "h264".into()
}

#[derive(Debug, Default, Deserialize)]
pub struct ClientInfo {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub os: String,
    #[serde(default)]
    pub app_version: String,
}

#[derive(Debug, Serialize)]
#[serde(tag = "type")]
pub enum ServerMsg {
    #[serde(rename = "welcome")]
    Welcome {
        session: String,
        server: ServerInfo,
        hmac: String,
        /// Long-lived token the client should persist and use in future
        /// `trusted_hello` messages — saves the user from scanning the QR
        /// every time. Only present on a successful first-time Hello,
        /// not on a TrustedHello reconnect (the client already has it).
        #[serde(skip_serializing_if = "Option::is_none")]
        trust_token: Option<String>,
        /// Stable identifier the server assigned to this device. The
        /// client echoes it back in `trusted_hello.device_id` so the
        /// server can look up the right token in O(1).
        #[serde(skip_serializing_if = "Option::is_none")]
        device_id: Option<String>,
    },
    #[serde(rename = "error")]
    Error { code: ErrorCode, msg: String },
    #[serde(rename = "pong")]
    Pong { ts: u64 },

    // ---- M2: stream control ----

    #[serde(rename = "stream_started")]
    StreamStarted {
        stream_id: String,
        codec: String,
        profile: String,
        level: String,
        width: u32,
        height: u32,
        fps: u32,
        bitrate_kbps: u32,
        keyframe_interval_frames: u32,
        pixel_format: String,
        started_at_unix_ms: u64,
        #[serde(skip_serializing_if = "Option::is_none")]
        audio: Option<AudioMetadata>,
    },

    #[serde(rename = "stream_stopped")]
    StreamStopped {
        stream_id: String,
        reason: StreamStopReason,
        msg: String,
    },

    /// Reply to a `clipboard_get` with the current PC clipboard text.
    #[serde(rename = "clipboard_text")]
    ClipboardText { text: String },

    // ---- M6: file transfer ----

    /// Server accepted the transfer and opened the destination file.
    /// Phone can start streaming chunks.
    #[serde(rename = "file_transfer_accepted")]
    FileTransferAccepted { id: u32, dest_path: String },

    /// Final ack — all chunks received, file closed and renamed into
    /// place. `dest_path` is the on-disk path (useful for the phone UI
    /// to show "已保存到 ...").
    #[serde(rename = "file_transfer_complete")]
    FileTransferComplete { id: u32, dest_path: String },

    /// Anything went wrong — bad name, no space, IO error, premature
    /// disconnect, abort, etc. Phone shows the reason to the user.
    #[serde(rename = "file_transfer_failed")]
    FileTransferFailed { id: u32, reason: String },

    // ---- M6 v2: file transfer (PC → phone). Mirrors `FileTransfer*`
    // but with the directions inverted. PC is the sender. ----

    /// PC announces an incoming file. Phone allocates a destination
    /// (app-private external storage in v1 so we don't need a runtime
    /// storage permission), opens the file, and replies with
    /// `FileSendAccepted`. PC then streams FILE Binary frames carrying
    /// the same `id`; the last chunk has the `LAST` flag set.
    #[serde(rename = "file_send_begin")]
    FileSendBegin {
        id: u32,
        name: String,
        size: u64,
    },
}

#[derive(Debug, Serialize)]
pub struct ServerInfo {
    pub name: String,
    pub os: String,
    pub version: String,
}

/// Audio sub-stream metadata embedded in [`ServerMsg::StreamStarted`] when the
/// server captures system audio alongside the screen.
#[derive(Debug, Serialize, Clone)]
pub struct AudioMetadata {
    pub codec: String,
    pub sample_rate: u32,
    pub channels: u32,
    pub frame_size_ms: u32,
    pub bitrate_kbps: u32,
    /// Base64 of Opus ID Header bytes (Android MediaCodec csd-0).
    pub csd_0_b64: String,
    /// Base64 of pre-skip nanoseconds as LE i64 (csd-1).
    pub csd_1_b64: String,
    /// Base64 of seek pre-roll nanoseconds as LE i64 (csd-2).
    pub csd_2_b64: String,
}

#[derive(Debug, Serialize, Clone, Copy)]
#[serde(rename_all = "snake_case")]
pub enum ErrorCode {
    BadPairingCode,
    CodeExpired,
    CodeUsed,
    VersionMismatch,
    Malformed,
    /// Reconnect path: device_id wasn't in `trusted_devices.json` (first
    /// time on a new server, or the file was wiped). Phone should fall
    /// back to QR pairing.
    UnknownDevice,
    /// Reconnect path: device_id known, but the supplied token didn't
    /// match. Either the token was rotated/revoked or it's a malicious
    /// client. Phone should clear its stored token and fall back to QR.
    BadTrustToken,
    // M2
    StreamUnavailable,
    StreamAlreadyRunning,
    NotAuthenticated,
    // M3
    #[allow(dead_code)]
    InputUnavailable,
    // M3.5
    #[allow(dead_code)]
    KeyEventInvalid,
}

#[derive(Debug, Serialize, Clone, Copy)]
#[serde(rename_all = "snake_case")]
pub enum StreamStopReason {
    ClientRequested,
    EncoderError,
    CaptureError,
    ServerShutdown,
}

// ============================================================================
// Binary video frame format (data plane, WebSocket Binary frame)
// ============================================================================
//
// See docs/PROTOCOL.md §"数据平面" for the wire layout. Constants here mirror
// the documented byte offsets so encode/decode stays in lock-step with the spec.

pub const FRAME_HEADER_LEN: usize = 12;

/// `frame_type` byte values
pub mod frame_type {
    pub const VIDEO: u8 = 0x01;
    #[allow(dead_code)]
    pub const AUDIO: u8 = 0x02; // reserved for M5
    /// M6 file-transfer chunk. Reuses the 12-byte header layout; the
    /// 8 bytes that hold `pts_us` for video are repurposed as
    /// `transfer_id (u32 LE) || chunk_seq (u32 LE)`.
    pub const FILE: u8 = 0x03;
}

/// `flags` byte bits
pub mod frame_flags {
    /// IDR / keyframe — the receiver may resume from this point.
    pub const KEYFRAME: u8 = 1 << 0;
    /// Payload is prefixed with parameter sets (SPS/PPS), inline. Always set on H.264 IDR frames.
    pub const CONFIG: u8 = 1 << 1;
    /// M6 file-transfer: marks the FINAL chunk in a transfer. After
    /// receiving a frame with this bit set, the server flushes + closes
    /// the destination file and emits `FileTransferComplete`.
    pub const LAST_CHUNK: u8 = 1 << 0;
}

/// Build a video frame: 12-byte header + Annex-B payload, ready for `WebSocket::send(Binary)`.
pub fn build_video_frame(payload: &[u8], pts_us: u64, keyframe: bool, has_config: bool) -> Vec<u8> {
    let mut out = Vec::with_capacity(FRAME_HEADER_LEN + payload.len());
    out.push(frame_type::VIDEO);
    let mut flags = 0u8;
    if keyframe {
        flags |= frame_flags::KEYFRAME;
    }
    if has_config {
        flags |= frame_flags::CONFIG;
    }
    out.push(flags);
    out.extend_from_slice(&[0u8, 0u8]); // reserved
    out.extend_from_slice(&pts_us.to_le_bytes());
    out.extend_from_slice(payload);
    out
}

/// Build an audio frame: 12-byte header (frame_type=AUDIO, flags=0) + Opus payload.
pub fn build_audio_frame(payload: &[u8], pts_us: u64) -> Vec<u8> {
    let mut out = Vec::with_capacity(FRAME_HEADER_LEN + payload.len());
    out.push(frame_type::AUDIO);
    out.push(0); // flags
    out.extend_from_slice(&[0u8, 0u8]); // reserved
    out.extend_from_slice(&pts_us.to_le_bytes());
    out.extend_from_slice(payload);
    out
}

/// Parse a binary frame's 12-byte header. Returns `(frame_type, flags,
/// pts_or_id_seq_bytes, payload_offset)`. Caller dispatches on
/// `frame_type` and re-interprets the 8-byte field accordingly.
pub fn peek_frame_header(buf: &[u8]) -> Option<(u8, u8)> {
    if buf.len() < FRAME_HEADER_LEN {
        return None;
    }
    Some((buf[0], buf[1]))
}

/// Build a FILE chunk frame for the PC → phone path (M6 v2). Mirror
/// of `parse_file_chunk`: the 8 PTS bytes carry `transfer_id (u32 LE)
/// || chunk_seq (u32 LE)`. `last_chunk` sets `frame_flags::LAST_CHUNK`
/// so the phone knows to finalize the destination file after writing
/// this payload.
pub fn build_file_chunk_frame(
    transfer_id: u32,
    chunk_seq: u32,
    last_chunk: bool,
    payload: &[u8],
) -> Vec<u8> {
    let mut out = Vec::with_capacity(FRAME_HEADER_LEN + payload.len());
    out.push(frame_type::FILE);
    out.push(if last_chunk { frame_flags::LAST_CHUNK } else { 0 });
    out.extend_from_slice(&[0u8, 0u8]); // reserved
    out.extend_from_slice(&transfer_id.to_le_bytes());
    out.extend_from_slice(&chunk_seq.to_le_bytes());
    out.extend_from_slice(payload);
    out
}

/// Decode a FILE chunk header (the 8 PTS bytes are split into
/// `transfer_id (u32 LE) || chunk_seq (u32 LE)`). Returns
/// `(transfer_id, chunk_seq, is_last, payload_slice)`.
pub fn parse_file_chunk(buf: &[u8]) -> Option<(u32, u32, bool, &[u8])> {
    if buf.len() < FRAME_HEADER_LEN {
        return None;
    }
    let flags = buf[1];
    let id = u32::from_le_bytes([buf[4], buf[5], buf[6], buf[7]]);
    let seq = u32::from_le_bytes([buf[8], buf[9], buf[10], buf[11]]);
    let is_last = (flags & frame_flags::LAST_CHUNK) != 0;
    Some((id, seq, is_last, &buf[FRAME_HEADER_LEN..]))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_video_frame_layout_matches_spec() {
        let payload = [0xAAu8, 0xBB, 0xCC];
        let frame = build_video_frame(&payload, 0x0102_0304_0506_0708, true, true);
        assert_eq!(frame.len(), FRAME_HEADER_LEN + payload.len());
        assert_eq!(frame[0], frame_type::VIDEO);
        assert_eq!(frame[1], frame_flags::KEYFRAME | frame_flags::CONFIG);
        assert_eq!(&frame[2..4], &[0, 0]);
        assert_eq!(&frame[4..12], &0x0102_0304_0506_0708u64.to_le_bytes());
        assert_eq!(&frame[12..], &payload);
    }

    #[test]
    fn p_frame_has_no_flags() {
        let frame = build_video_frame(&[0u8; 8], 1234, false, false);
        assert_eq!(frame[1], 0);
    }
}
